gtk/Session.cc

```diff
@@ -1222,29 +1222,35 @@ void Session::Impl::send_rpc_request(
 
 void Session::port_test()
 {
-    auto const tag = nextTag;
-    ++nextTag;
+    auto const response_func = [this](tr_variant& response)
+    {
+        tr_variant* args = nullptr;
+        auto is_open = false;
+        auto ip_protocol = std::string_view{};
 
-    tr_variant request;
-    tr_variantInitDict(&request, 2);
-    tr_variantDictAddStrView(&request, TR_KEY_method, "port-test");
-    tr_variantDictAddInt(&request, TR_KEY_tag, tag);
-    impl_->send_rpc_request(
-        request,
-        tag,
-        [this](auto& response)
+        if (tr_variantDictFindDict(&response, TR_KEY_arguments, &args))
         {
-            tr_variant* args = nullptr;
-            bool is_open = false;
+            tr_variantDictFindBool(args, TR_KEY_port_is_open, &is_open);
+            tr_variantDictFindStrView(args, TR_KEY_ipProtocol, &ip_protocol);
+        }
 
-            if (!tr_variantDictFindDict(&response, TR_KEY_arguments, &args) ||
-                !tr_variantDictFindBool(args, TR_KEY_port_is_open, &is_open))
-            {
-                is_open = false;
-            }
+        impl_->signal_port_tested().emit(ip_protocol, is_open);
```

Comment:
One call is made to `Session` to test a port but two signals are emitted back, and then `PrefsDialog` expects exactly two or it won't function right — feels like `PrefsDialog` relies on `Session` internals; I'd rather either have `PrefsDialog` make two calls and expect two signals, or have `Session` emit only one signal in response to one call, e.g. `port_tested(bool is_ipv4_open, bool is_ipv6_open)`.

Reply:
fd8dd0a GTK
a5aa955 Qt

> or have `Session` emit only one signal in response to one call, e.g. `port_tested(bool is_ipv4_open, bool is_ipv6_open)`.

I'd like it to behave this way too, especially if we are only going to use 1 label. It was quite tricky though, see if you are happy with it.